flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/checkpoints/TaskCheckpointStatisticDetailsHandler.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.flink.runtime.rest.messages.checkpoints.MinMaxAvgStatistics;
+import org.apache.flink.runtime.rest.messages.checkpoints.SubtaskCheckpointStatistics;
+import org.apache.flink.runtime.rest.messages.checkpoints.TaskCheckpointMessageParameters;
+import org.apache.flink.runtime.rest.messages.checkpoints.TaskCheckpointStatisticsWithSubtaskDetails;
+import org.apache.flink.runtime.webmonitor.RestfulGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+/**
+ * REST handler which serves checkpoint statistics for subtasks.
+ */
+public class TaskCheckpointStatisticDetailsHandler extends AbstractCheckpointHandler<TaskCheckpointStatisticsWithSubtaskDetails, TaskCheckpointMessageParameters> {
+
+	public TaskCheckpointStatisticDetailsHandler(
+			CompletableFuture<String> localRestAddress,
+			GatewayRetriever<? extends RestfulGateway> leaderRetriever,
+			Time timeout,
+			MessageHeaders<EmptyRequestBody, TaskCheckpointStatisticsWithSubtaskDetails, TaskCheckpointMessageParameters> messageHeaders,
+			ExecutionGraphCache executionGraphCache,
+			Executor executor,
+			CheckpointStatsCache checkpointStatsCache) {
+		super(localRestAddress, leaderRetriever, timeout, messageHeaders, executionGraphCache, executor, checkpointStatsCache);
+	}
+
+	@Override
+	protected TaskCheckpointStatisticsWithSubtaskDetails handleCheckpointRequest(
+			HandlerRequest<EmptyRequestBody, TaskCheckpointMessageParameters> request,
+			AbstractCheckpointStats checkpointStats) throws RestHandlerException {
+
+		final JobVertexID jobVertexId = request.getPathParameter(JobVertexIdPathParameter.class);
+
+		final TaskStateStats taskStatistics = checkpointStats.getTaskStateStats(jobVertexId);
+
+		if (taskStatistics != null) {
+
+			final TaskCheckpointStatisticsWithSubtaskDetails.Summary summary = createSummary(
+				taskStatistics.getSummaryStats(),
+				checkpointStats.getTriggerTimestamp());
+
+			final List<SubtaskCheckpointStatistics> subtaskCheckpointStatistics = createSubtaskCheckpointStatistics(
+				taskStatistics.getSubtaskStats(),
+				checkpointStats.getTriggerTimestamp());
+
+			return new TaskCheckpointStatisticsWithSubtaskDetails(
+				checkpointStats.getCheckpointId(),
+				checkpointStats.getStatus(),
+				taskStatistics.getLatestAckTimestamp(),
+				taskStatistics.getStateSize(),
+				taskStatistics.getEndToEndDuration(checkpointStats.getTriggerTimestamp()),
+				taskStatistics.getAlignmentBuffered(),
+				taskStatistics.getNumberOfSubtasks(),
+				taskStatistics.getNumberOfAcknowledgedSubtasks(),
+				summary,
+				subtaskCheckpointStatistics);
+		} else {
+			throw new RestHandlerException("There is no checkpoint statistics for task " + jobVertexId + '.', HttpResponseStatus.NOT_FOUND);
```

Comment:
"There are"?

Reply:
I think statistics should be used with a singular verb, but I'm not entirely sure, either.